quixstreams/rowconsumer.py

```diff
@@ -149,4 +158,77 @@ def poll_row(self, timeout: Optional[float] = None) -> Union[Row, List[Row], Non
 
     def close(self):
         super().close()
+        self.reset_backpressure()
         self._inner_consumer = None
+
+    @property
+    def backpressured_tps(self) -> set[TopicPartition]:
+        return self._backpressurred_tps
+
+    def trigger_backpressure(
+        self,
+        offsets_to_seek: dict[tuple[str, int], int],
+        resume_after: float,
+    ):
+        """
+        Pause all partitions for the certain period of time and seek the partitions
+        provided in the `offsets_to_seek` dict.
+
+        This method is supposed to be called in case of backpressure from Sinks.
+        """
+        resume_at = monotonic() + resume_after
+        self._backpressure_resume_at = min(self._backpressure_resume_at, resume_at)
+
+        # Pause only data TPs excluding changelog TPs
+        non_changelog_topics = {
+            k: v for k, v in self._topics.items() if not v.is_changelog
+        }
+        non_changelog_tps = [
+            tp for tp in self.assignment() if tp.topic in non_changelog_topics
+        ]
+
+        for tp in non_changelog_tps:
+            position, *_ = self.position([tp])
+            logger.debug(
+                f'Pausing topic partition "{tp.topic}[{tp.partition}]" for {resume_after}s; '
+                f"position={position.offset}"
+            )
+            self.pause(partitions=[tp])
+
+            # Seek the TP back to the "offset_to_seek" to start from it on resume.
+            # The "offset_to_seek" is provided by the Checkpoint and is expected to be the
+            # first offset processed in the checkpoint.
+            # There may be no offset for the TP if no message has been processed yet.
+            seek_offset = offsets_to_seek.get((tp.topic, tp.partition))
+            if seek_offset is not None:
+                logger.debug(
+                    f'Seek the paused partition "{tp.topic}[{tp.partition}]" back to '
+                    f"offset {seek_offset}"
+                )
+                self.seek(
+                    partition=TopicPartition(
+                        topic=tp.topic, partition=tp.partition, offset=seek_offset
+                    )
+                )
+
+            self._backpressurred_tps.add(tp)
+
+    def resume_backpressured(self):
+        """
+        Resume consuming from assigned data partitions after the wait period has elapsed.
+        """
+        if self._backpressure_resume_at > monotonic():
+            return
+
+        # Resume the previously backpressured TPs
+        for tp in self._backpressurred_tps:
+            logger.debug(f'Resuming topic partition "{tp.topic}[{tp.partition}]"')
+            self.resume(
+                partitions=[TopicPartition(topic=tp.topic, partition=tp.partition)]
+            )
```

Comment:
Is there a reason we don't want to pass the existing tp object?
```suggestion
            self.resume(partitions=[tp])
```

Reply:
Not really. 
The original `TopicPartition` object may have an additional attribute set (e.g. `offset`), but `.resume()` should ignore them.

I'll update that